### PR TITLE
Filter Operations and Conventions

### PR DESCRIPTION
**Pull request by PascalSenn**
Adds conventions and operations

### CODE REVIEW

File: src/HotChocolate/Core/src/Types/Types/Descriptors/ObjectFieldDescriptor.cs [hunk 1]
```diff
@@ -16,7 +16,7 @@ public class ObjectFieldDescriptor
     {
         private bool _argumentsInitialized;
 
-        protected ObjectFieldDescriptor(
+        protected internal ObjectFieldDescriptor(
```
Comment (michaelstaib) [inline src/HotChocolate/Core/src/Types/Types/Descriptors/ObjectFieldDescriptor.cs:19]: I just deliberately removed this. Why do you not use the factory methods?

File: src/HotChocolate/Data/src/Data/Filters/Convention/FilterConvention.cs [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+            IConventionContext context,
+            FilterConventionDefinition? definition)
+        {
+            Operations = definition
+                .Operations
+                .ToImmutableDictionary(x => x.Operation, x => new OperationConvention(x));
+        }
+
+        public NameString GetFieldDescription(IDescriptorContext context, MemberInfo member)
+            => context.Naming.GetMemberDescription(member, MemberKind.InputObjectField);
```
Comment (michaelstaib) [inline src/HotChocolate/Data/src/Data/Filters/Convention/FilterConvention.cs:53]: we have to find one style on this.

A)

```csharp
public NameString GetFieldDescription(IDescriptorContext context, MemberInfo member)
    => context.Naming.GetMemberDescription(member, MemberKind.InputObjectField);
````

B)

```csharp
public NameString GetFieldDescription(IDescriptorContext context, MemberInfo member) => 
    context.Naming.GetMemberDescription(member, MemberKind.InputObjectField);
````

@rstaib @PascalSenn what is each once preference?

File: src/HotChocolate/Data/src/Data/Filters/Convention/FilterConvention.cs [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+            return null;
+        }
+
+        public NameString GetOperationName(IDescriptorContext context, int operation)
+        {
+            if (Operations.TryGetValue(operation, out OperationConvention? operationConvention))
+            {
+                return operationConvention.Name;
+            }
+            throw new SchemaException(
```
Comment (michaelstaib) [inline src/HotChocolate/Data/src/Data/Filters/Convention/FilterConvention.cs:76]: All exceptions (apart from argument exceptions) need to be moved to a ThrowHelper... like in execution.

File: src/HotChocolate/Data/src/Data/Filters/Convention/FilterConventionDefinition.cs [hunk 4]
```diff
@@ -0,0 +1,13 @@
+using System.Collections.Generic;
+using System.Linq;
+
+namespace HotChocolate.Data.Filters
+{
+    public class FilterConventionDefinition
+    {
+        public IEnumerable<FilterOperationConventionDefinition> Operations { get; set; } =
```
Comment (michaelstaib) [inline src/HotChocolate/Data/src/Data/Filters/Convention/FilterConventionDefinition.cs:8]: Also here I see 2 styles, which are we gonna choose:

A)

```csharp
public IEnumerable<FilterOperationConventionDefinition> Operations { get; set; } =
    Enumerable.Empty<FilterOperationConventionDefinition>();
```

B)

```csharp
public IEnumerable<FilterOperationConventionDefinition> Operations { get; set; } 
    = Enumerable.Empty<FilterOperationConventionDefinition>();
```

@rstaib @PascalSenn

File: src/HotChocolate/Data/src/Data/Filters/Convention/FilterConvention.cs [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+
+        protected FilterConvention()
+        {
+            _configure = Configure;
+        }
+
+        public FilterConvention(Action<IFilterConventionDescriptor> configure)
+        {
+            _configure = configure
+                ?? throw new ArgumentNullException(nameof(configure));
```
Comment (michaelstaib) [inline src/HotChocolate/Data/src/Data/Filters/Convention/FilterConvention.cs:27]: Also here we have now 2 styles, which are we gonna choose?

A)

```csharp
_configure = configure
    ?? throw new ArgumentNullException(nameof(configure));
```

B)

```csharp
_configure = configure ?? 
    throw new ArgumentNullException(nameof(configure));
```

@rstaib @PascalSenn

Comment (PascalSenn) [reply src/HotChocolate/Data/src/Data/Filters/Convention/FilterConvention.cs:27]: B)

File: src/HotChocolate/Data/src/Data/Filters/Convention/FilterConventionDefinition.cs (hunk 4 above)
Comment (PascalSenn) [reply src/HotChocolate/Data/src/Data/Filters/Convention/FilterConventionDefinition.cs:8]: A)

File: src/HotChocolate/Data/src/Data/Filters/Convention/FilterConvention.cs (hunk 2 above)
Comment (PascalSenn) [reply src/HotChocolate/Data/src/Data/Filters/Convention/FilterConvention.cs:53]: B)

Comment (rstaib) [reply src/HotChocolate/Data/src/Data/Filters/Convention/FilterConvention.cs:53]: B)

File: src/HotChocolate/Data/src/Data/Filters/Convention/FilterConventionDefinition.cs (hunk 4 above)
Comment (rstaib) [reply src/HotChocolate/Data/src/Data/Filters/Convention/FilterConventionDefinition.cs:8]: A)

File: src/HotChocolate/Data/src/Data/Filters/Convention/FilterConvention.cs (hunk 5 above)
Comment (rstaib) [reply src/HotChocolate/Data/src/Data/Filters/Convention/FilterConvention.cs:27]: B)

**Pull request merged**